File: src/Exchange/Main/MatchingEngine/Limit.java
package Exchange.Main.MatchingEngine;

import java.math.BigDecimal;
import java.util.Iterator;
import java.util.NoSuchElementException;

// LinkedList
// Tail.getNext = null always
public class Limit implements Iterable<Order>, Iterator<Order> {
    private BigDecimal LimitPrice;
    private Integer Volume; // Number of outstanding shares
    private Integer Length; // Number of outstanding orders

    private Order HeadOrder;
    private Order TailOrder;
    private Order Last; // last = last element that was returned by an iterator-returning-function

    public Limit(BigDecimal LimitPrice){
        this.LimitPrice = LimitPrice;
        Volume = 0;
        Length = 0;

        HeadOrder = null;
        TailOrder = null;
        Last = null;
    }

    @Override
    public Iterator<Order> iterator() {
        this.Last = HeadOrder;
        return this;
    }

    @Override
    public boolean hasNext() {
        return this.Last != null;
    }

    @Override
    public Order next() {
        if (this.Last == null) {
            throw new NoSuchElementException();
        }
        Order returnVal = this.Last;
        this.Last = this.Last.getNextOrder();
        return returnVal;
    }
    public void remove() {
        throw new UnsupportedOperationException();
    }

    public void appendOrder(Order incomingOrder){
        incomingOrder.setParentLimit(this);
        if (Length.equals(0)) {
            incomingOrder.setNextOrder(null);
            incomingOrder.setPreviousOrder(null);
            HeadOrder = incomingOrder;
            TailOrder = incomingOrder;
        } else{
            incomingOrder.setPreviousOrder(TailOrder);
            incomingOrder.setNextOrder(null);
            TailOrder.setNextOrder(incomingOrder); // LastAddedOrder.setNext(incomingOrder)
            TailOrder = incomingOrder;
        }
        Length+=1;
        Volume += incomingOrder.getQuantity();
    }
    public void removeOrder(Order order) throws NoSuchElementException{
        if (this.Length == 0) {
            throw new NoSuchElementException("Order Doesn't Exist");
        }
        Order tempNextOrder = order.getNextOrder();
        Order tempPrevOrder = order.getPreviousOrder();
        if ((tempNextOrder != null) && (tempPrevOrder != null)) {
            tempNextOrder.setPreviousOrder(tempPrevOrder);
            tempPrevOrder.setNextOrder(tempNextOrder);
        } else if (tempNextOrder != null){
            tempNextOrder.setPreviousOrder(null);
            this.HeadOrder = tempNextOrder;
        } else if (tempPrevOrder != null){
            tempPrevOrder.setNextOrder(null);
            this.TailOrder = tempPrevOrder;
        }
        this.Volume -= order.getQuantity();
        this.Length -= 1;
    }
    public void movetoTail(Order order) {
        /*
         * Move 'order' to the tail of the list (after modification for e.g.)
         */
        if (order.getPreviousOrder() != null) {
            order.getPreviousOrder().setNextOrder(order.getNextOrder());
        } else {
            // Update head order
            this.HeadOrder = order.getNextOrder();
        }
        order.getNextOrder().setPreviousOrder(order.getPreviousOrder());
        // Set the previous tail's next order to this order
        this.TailOrder.setNextOrder(order);
        order.setPreviousOrder(this.TailOrder);
        this.TailOrder = order;
        order.setNextOrder(null);
    }
    public String toString() {
        StringBuilder outString = new StringBuilder();
        for (Order o : this) {
            outString.append("| ").append(o.toString()).append("\n");
        }
        return outString.toString();
    }

    public Integer getVolume() {
        return Volume;
    }

    public Integer getLength() {
        return Length;
    }

    public BigDecimal getLimitPrice() {
        return LimitPrice;
    }

    public Order getHeadOrder() {
        return HeadOrder;
    }

    public Order getTailOrder() {
        return TailOrder;
    }

    public void setVolume(Integer volume) {
        Volume = volume;
    }
}

File: src/GAMFinancial-WebApp/GAMFinancial-WebApp/Models/Limit.cs
using java.lang;
using java.util;
using java.util.function;
using System.Collections;

namespace GAMFinancial_WebApp.Models
{
    public class Limit
    {
        public Limit(decimal limitPrice)
        {
            this.LimitPrice = limitPrice;
            this.Volume = 0;
            this.Length = 0;

            this.HeadOrder = null;
            this.TailOrder = null;
        }

        public int Id { get; set; }

        public decimal LimitPrice { get; set; }

        public int Length { get; set; }

        public int Volume { get; set; }

        public Order? HeadOrder { get; set; }

        public Order? TailOrder { get; set; }

        public void appendOrder(Order incomingOrder)
        {
            incomingOrder.ParentLimit = this;
            if (Length.Equals(0))
            {
                incomingOrder.NextOrder = null;
                incomingOrder.PreviousOrder = null;
                HeadOrder = incomingOrder;
                TailOrder = incomingOrder;
            }
            else
            {
                incomingOrder.PreviousOrder = this.TailOrder;
                incomingOrder.NextOrder = null;
                this.TailOrder.NextOrder = incomingOrder; // LastAddedOrder.setNext(incomingOrder)
                TailOrder = incomingOrder;
            }
            Length+=1;
            Volume += incomingOrder.Quantity;
        }
        public void removeOrder(Order order)
        {
            if (this.Length == 0) 
            {
                throw new NoSuchElementException("Order Doesn't Exist");
            }

            Order tempNextOrder = order.NextOrder;
            Order tempPrevOrder = order.PreviousOrder;
            if ((tempNextOrder != null) && (tempPrevOrder != null)) 
            {
                tempNextOrder.PreviousOrder = tempPrevOrder;
                tempPrevOrder.NextOrder = tempNextOrder;
            } 
            else if (tempNextOrder != null)
            {
                    tempNextOrder.PreviousOrder = null;
                    this.HeadOrder = tempNextOrder;
            } 
            else if (tempPrevOrder != null)
            {
                tempPrevOrder.NextOrder = null;
                this.TailOrder = tempPrevOrder;
            }
            this.Volume -= order.Quantity;
            this.Length -= 1;
        }
        public void movetoTail(Order order)
        {
            /*
             * Move 'order' to the tail of the list (after modification for e.g.)
             */
            if (order.PreviousOrder != null)
            {
                order.PreviousOrder.NextOrder = order.NextOrder;
            }
            else
            {
                // Update head order
                this.HeadOrder = order.NextOrder;
            }
            order.NextOrder.PreviousOrder = order.PreviousOrder;
            // Set the previous tail's next order to this order
            this.TailOrder.NextOrder = order;
            order.PreviousOrder = this.TailOrder;
            this.TailOrder = order;
            order.NextOrder = null;
        }
    }
}
